dask/array/core.py

```diff
@@ -1485,28 +1485,12 @@ def __rxor__(self, other):
         return elemwise(operator.xor, other, self)
 
     def __matmul__(self, other):
-        from .routines import tensordot
-        other = asanyarray(other)
-        if other.ndim > 2:
-            msg = ('The matrix multiplication operator (@) is not yet '
-                   'implemented for higher-dimensional Dask arrays. Try '
-                   '`dask.array.tensordot` and see the discussion at '
-                   'https://github.com/dask/dask/pull/2349 for details.')
-            raise NotImplementedError(msg)
-        return tensordot(self, other, axes=((self.ndim - 1,),
-                                            (other.ndim - 2,)))
+        from .routines import matmul
+        return matmul(self, other)
```

Comment:
It would be nice if `dask.array` deferred to other types when appropriate (by returning `NotImplemented`), but that's a bigger issue than for `__matmul__`.

Reply:
In any event, this should no longer be an issue for `matmul` given these changes (unless I'm missing something).

Reply:
`matmul` casts inputs with `asanyarray` and as far as I can tell never returns `NotImplemented`. So there isn't any way to write another object that overrides arithmetic with dask arrays.

I'll open another issue for this.

Reply:
Ok, I think I misunderstood you before.

So my understanding of `da.asanyarray` was it should leave anything that inherits from a Dask Array untouched. Does that work for the use cases you have in mind? If not, are we wanting to extend functions like `da.matmul` to objects that do not inherit from a Dask Array? If so, how do we expect things like `da.atop` to behave in these contexts?

Reply:
I don't think we should extend `da.matmul` to handle other non-dask array types.

We should just be aware that there are cases where `dask_array @ other_obj` should be implemented by `other_obj.__rmatmul__` rather than `dask_array.__matmul__`. But to be clear, I don't think this should block this particular PR, since this is a much bigger issue (https://github.com/dask/dask/issues/2913).

Reply:
Sure. I agree. Not totally sure how to fix it yet. Let's see how the discussion in ( https://github.com/dask/dask/issues/2913 ) unwinds.